pkg/controllers/monitoring/monitoring-stack/controller.go

```diff
@@ -80,32 +85,35 @@ func RegisterWithManager(mgr ctrl.Manager, opts Options) error {
 		return fmt.Errorf("invalid InstanceSelector: %s", opts.InstanceSelector)
 	}
 
+	mcClient, err := monitoringclient.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return err
+	}
+
+	conditions := CreateConditions()
+
 	r := &reconciler{
 		k8sClient:             mgr.GetClient(),
+		mcClient:              mcClient,
 		scheme:                mgr.GetScheme(),
 		logger:                ctrl.Log.WithName("observability-operator"),
 		instanceSelectorKey:   split[0],
 		instanceSelectorValue: split[1],
 		grafanaDSWatchCreated: false,
+		conditions:            conditions,
 	}
 
-	// We only want to trigger a reconciliation when the generation
-	// of a child changes. Until we need to update our the status for our own objects,
-	// we can save CPU cycles by avoiding reconciliations triggered by
-	// child status changes.
-	p := predicate.GenerationChangedPredicate{}
-
 	ctrl, err := ctrl.NewControllerManagedBy(mgr).
 		WithLogger(ctrl.Log).
-		For(&stack.MonitoringStack{}).
-		Owns(&monv1.Prometheus{}).WithEventFilter(p).
-		Owns(&monv1.Alertmanager{}).WithEventFilter(p).
-		Owns(&v1.Service{}).WithEventFilter(p).
-		Owns(&v1.ServiceAccount{}).WithEventFilter(p).
-		Owns(&rbacv1.Role{}).WithEventFilter(p).
-		Owns(&rbacv1.RoleBinding{}).WithEventFilter(p).
-		Owns(&monv1.ServiceMonitor{}).WithEventFilter(p).
-		Owns(&policyv1.PodDisruptionBudget{}).WithEventFilter(p).
+		For(&stack.MonitoringStack{}).WithEventFilter(predicate.ResourceVersionChangedPredicate{}).
+		Owns(&monv1.Prometheus{}).
+		Owns(&monv1.Alertmanager{}).
+		Owns(&v1.Service{}).
+		Owns(&v1.ServiceAccount{}).
+		Owns(&rbacv1.Role{}).
+		Owns(&rbacv1.RoleBinding{}).
+		Owns(&monv1.ServiceMonitor{}).
+		Owns(&policyv1.PodDisruptionBudget{}).
```

Comment:
TBH I don't know how these filter predicates are supposed to work when there are more of them. So it would be good if some k8s controller-runtime expert double check. 
I was struggling (with original definition) with Prometheus being not available and no additional reconcile attempt (so the conditions were not updated correctly)

Reply:
I think you're on something. IIUC `*Builder.WithEventFilter()` applies to all watched objects. We need [Predicates.ApplyToOwns](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.2/pkg/builder#Predicates.ApplyToOwns) for the Prometheus resource:

```golang
Owns(&monv1.Prometheus{}, builder.WithPredicates(predicate.ResourceVersionChangedPredicate{}))
```

Reply:
Thanks @simonpasquier. It seems to work fine. Let me update the PR.